Sources/NIOSSL/NIOSSLHandler.swift

```diff
@@ -914,42 +928,60 @@ extension NIOSSLHandler {
-                        _ = bufferedActions.removeFirst()
+            var bufferedActionsLoopCount = 0
+            bufferedActionsLoop: while self.bufferedActions.hasMark {
+                bufferedActionsLoopCount += 1
+                var didWrite = false
+
+                writeLoop: while self.bufferedActions.hasMark {
+                    let element = self.bufferedActions.first!
+                    switch element {
+                    case .write(let bufferedWrite):
+                        var data = bufferedWrite.data
+                        let writeSuccessful = try self._encodeSingleWrite(buf: &data)
+                        if writeSuccessful {
+                            didWrite = true
+                            if let promise = bufferedWrite.promise { promises.append(promise) }
+                            _ = self.bufferedActions.removeFirst()
+                        } else {
+                            // The write into BoringSSL unsuccessful. Break the write loop so any
+                            // data is written to the network before resuming.
+                            break writeLoop
+                        }
+                    case .closeOutput:
+                        invokeCloseOutput = true
+                        _ = self.bufferedActions.removeFirst()
+                        break writeLoop
                     }
-                case .closeOutput:
-                    invokeCloseOutput = true
-                    _ = bufferedActions.removeFirst()
+                }
+
+                // If we got this far and did a write, we should shove the data out to the
+                // network.
+                if didWrite {
+                    let ourPromise: EventLoopPromise<Void>? = promises.flattenPromises(on: context.eventLoop)
+                    self.writeDataToNetwork(context: context, promise: ourPromise)
+                }
+
+                // We detected a .closeOutput action in our action buffer. This means we
+                // close the output after we have written all pending writes.
+                if invokeCloseOutput {
+                    self.state = .outputClosed
+                    self.doShutdownStep(context: context)
+                    self.discardBufferedActions(reason: ChannelError.outputClosed)
                     break bufferedActionsLoop
                 }
-            }
 
-            // If we got this far and did a write, we should shove the data out to the
-            // network.
-            if didWrite {
-                let ourPromise: EventLoopPromise<Void>? = promises.flattenPromises(on: context.eventLoop)
-                self.writeDataToNetwork(context: context, promise: ourPromise)
+                // We spun the outer loop too many times, something isn't right so let's bail out
+                // instead of looping any longer.
+                if bufferedActionsLoopCount >= 1000 {
```

Comment:
Should this check be at the top?

Reply:
Do you mean as part of `bufferedActionsLoop: while self.bufferedActions.hasMark {`? i.e.

```swift
bufferedActionsLoop: while self.bufferedActions.hasMark && bufferedActionsLoopCount < 1000 {
```

Reply:
Yeah